### ISSUE TITLE
Start timely with dynamic hosts. 

### ISSUE THREAD

**Issue by bmmcq** [frankmcsherry/timely-dataflow#182]
Hi, @frankmcsherry,

In my application, every server's IP and port are hard to determinate from the beginning. So I wonder if it's possible to start "timely" with dynamic hosts which will be changed at runtime. 

Below is my idea: 

![image](https://user-images.githubusercontent.com/32241292/43201955-97c6ea7c-904c-11e8-82f4-e1b51d991fb3.png)

A "Center" can be some data persistent services, which is provided by the user. All "timely" need to do is to provide a "Center" trait. Maybe like the following : 

```rust 
pub trait Center {
    fn register(&self, id : u64, address : String) -> Result<Vec<String>, String>;
}

```

So, if the user provides a 'Center' trait implementation, 'timely' process can bind on a random port first, start to await connections, and then use the "Center" to register itself, fetch back other processes' address. Finally, use the fetched address to start connecting ...

I'm looking forward to your suggestion. 
Thanks!

**Comment by frankmcsherry** [comment]
This doesn't seem too hard to do, and sounds like it could be helpful. 

The specific bit of code we'll have to check out is in [initialize_networking](https://github.com/frankmcsherry/timely-dataflow/blob/master/communication/src/networking.rs#L234-L235), where we start two threads to create connections to others and to await connections from others. If we have a bit more discipline here, we can set up a connection first, then call your `register` method in `Center`, and then start connections to whatever it gets back.

It seems like we'll need to put a bit of the communication set-up up on bricks, but it should be pretty easy to prototype up. We can probably just hack a bit around [create_allocators](https://github.com/frankmcsherry/timely-dataflow/blob/master/communication/src/initialize.rs#L76-L89) and let you tap in there.

Rather than a `Center` trait, can we just go with something like a 
```rust
FnOnce(u64, String)->Result<Vec<String>,String>
```

**Comment by frankmcsherry** [comment]
Ok, looking more at the code and thinking alternately:

You should have the ability to build your own `communication::Configuration::Cluster` instance, which has a list of addresses and ports in it. I *suspect* that you could build a trampoline outside the current networking code that just builds up this object carefully.

Bear with me, and jump in if you think this is wrong:

If when starting up, before you initiate a timely dataflow computation, your first action is to call whatever you have behind `register`, which heads off to the `Center` and returns with a `Vec<String>` once it has enough registrations. Once it returns, you can construct a configuration like so:

```rust
if let Ok(addresses) = center.register(process_id, local_address) {
    let configuration = Configuration::Cluster(local_threads, process_id, addresses, report);
    timely.initialize(configuration, log_stuff, |worker| {

    });
}
```

**Comment by bmmcq** [comment]
Thank you very much for your prompt response.
I think it's fine to use "FnOnce" instead of a trait.
But the second code you suggested may have a little problem. 
Because when the process registering itself with IP and a random port,  it must already bind on that port and waiting for new connections, otherwise the random port may be preempted by other processes on the same machine. 
I have tried to modify a bit of code in the “initialize_networking”, as follows: 

```rust
/// Initializes network connections
pub fn initialize_networking<C:Center>(
    addresses: Vec<String>, my_index: usize, threads: usize, noisy: bool, center: Option<&C>, log_sender: Arc<Fn(::logging::CommsSetup)->::logging::CommsLogger+Send+Sync>) -> Result<Vec<Binary>> {

    let processes = addresses.len(); // The processes should always equal to the number of processes user configured.
    let hosts1;
    let await_task;

    if let Some(ref c) = center {
        let listener = try!(TcpListener::bind("0.0.0.0:0"));
        let addr = try!(listener.local_addr());

        // start a new thread to await connection.
        await_task = thread::spawn(move || {
            let mut results: Vec<_> = (0..(processes - my_index - 1)).map(|_| None).collect();
            for _ in (my_index + 1) .. processes {
                let mut stream = try!(listener.accept()).0;
                stream.set_nodelay(true).expect("set_nodelay call failed");
                let identifier = try!(stream.read_u64::<LittleEndian>()) as usize;
                results[identifier - my_index - 1] = Some(stream);
                if noisy { println!("worker {}:\tconnection from worker {}", my_index, identifier); }
            }
        });

        let my_address = format!("{}:{}", addr.ip(), addr.port());
        hosts1 = Arc::new(try!(c.register(my_index, my_address)));
    } else {
        hosts1 = Arc::new(addresses);
        await_task = thread::spawn(move || await_connections(hosts1, my_index, noisy));
    }
    ...
}

```

**Comment by frankmcsherry** [comment]
I see. I hadn't realized the dynamic port assignment was an important feature. Perhaps a simpler thing to do is for me to separate the `initialize_networking` method into two parts: i. acquire connections to peers (however you want to do this), and ii. build the timely-specific stuff around those connections.

This would let you plug in arbitrary logic for how to establish connections, perhaps first sniff around for open ports, then chat with a coordinator, then start up the connections the coordinator indicates. Timely doesn't really need to have an opinion on how this happens, other than to provide a default implementation that looks like what it does now.

**Comment by frankmcsherry** [comment]
Also, as a heads up on likely reality: this code is currently undergoing a rework in the `dataplane_mockup` branch, where the interaction between the networking threads and the rest of the system is being redone. My guess is that while we can try to land something that works for you with or after that change, there isn't likely to be much durable work done before then. 

It sounds like the main pain point for you is needing dynamically assigned ports, and it seems like this is pretty easy to hack around in a local copy of the repo. I'll try to make sure there is a clean separation between connection set-up and what happens with them in the new branch, and it shouldn't be too hard to make sure that you get the ability to tap in there.

**Comment by frankmcsherry** [comment]
The [dataplane_mockup](https://github.com/frankmcsherry/timely-dataflow/pull/135) branch should be landing soon, and exposes a new method `initialize_from`, which is meant to alleviate the problem you have. It allows you to provide your own vector of `A: AllocateBuilder` from which we create allocators. In principle, you can use / copy_paste the existing patterns and build either of the networked allocators (I recommend the newest) ahead of time, using the dynamic port assignment. Then you just call `initialize_from` with them as done by timely and you should be good to go (with a lot more boiler plate than before, though).

Timely doesn't help you do this, but this *should* allow you to do what you need to, with a small amount of pain. If we learn more about how to reduce this pain we can move that in to timely too!

**Comment by frankmcsherry** [comment]
This should now be available in #135. You can use timely's `execute::execute_from` to start a computation with a pre-build set of `Vec<A: AllocatorBuilder>` and you *should* be able to use the parts from communication to build a networked communicator on your own (that is, with careful staging to get the right port numbers, once listeners are bound).

If you have any questions on how to do this, please let me know.

**Comment by bmmcq** [comment]
Thanks very much, I can't wait to have a try.
